midend/checkSize.h

```diff
@@ -0,0 +1,46 @@
+/*
+Copyright 2018 VMware, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef _MIDEND_CHECKSIZE_H_
+#define _MIDEND_CHECKSIZE_H_
+
+#include "ir/ir.h"
+
+namespace P4 {
+
+/// Checks some possible misuses of the table size property
+class CheckTableSize : public Inspector {
```

Comment:
Would it make sense to not make this an inspector and actually replace the size by 0 in that case, so that the backend doesn't have to worry about it?

Reply:
Why not just remove the size property rather than replace it by 0?

Reply:
I don't really have a preference. It seems that either way each backend has to be careful. Currently the bmv2 backend unconditionally sets the table size to 1024 when the size is missing (or when the provided size is 0), which seems wrong: https://github.com/p4lang/p4c/blob/master/backends/bmv2/common/control.cpp#L386

Reply:
I think that this behavior is inherited from the p4-14 compiler.
The back-end should probably do this only when it is really necessary - i.e., the table size can change dynamically.

Reply:
bmv2 actually enforced fixed table sizes to emulate hardware targets
the default table size code should probably be updated to account for the key size, at least for exact match tables?

Reply:
Then let's file a new issue. Probably @hanw is the right person to handle this, since I believe he has ownership of the bmv2-based backends. For constant tables the compiler should somehow measure the table size and tell bmv2 what that is.

Reply:
In the meantime it looks to me like deleting the table size for these two cases is the right thing. The back-end is supposed to measure the table anyway. 0 seems a bit arbitrary.